lib/axon/loop.ex

```diff
@@ -968,6 +968,95 @@ defmodule Axon.Loop do
 
   defp default_checkpoint_file(epoch, _iter), do: "checkpoint_#{epoch}.ckpt"
 
+  @doc """
+  Adds a handler function which halts a loop if the given
+  metric does not improve between events.
+
+  By default, this will run after each epoch and track the
+  improvement of a given metric.
+
+  You must specify a metric to monitor and the metric must
+  be present in the loop state. Typically, this will be
+  a validation metric:
+
+      model
+      |> Axon.Loop.trainer(loss, optim)
+      |> Axon.Loop.metric(:accuracy)
+      |> Axon.Loop.validate(val_data)
+      |> Axon.Loop.early_stop("validation_accuracy")
+
+  It's important to remember that handlers are executed in the
+  order they are added to the loop. For example, if you'd like
+  to checkpoint a loop after every epoch and use early stopping,
+  most likely you want to add the checkpoint handler before
+  the early stopping handler:
+
+      model
+      |> Axon.Loop.trainer(loss, optim)
+      |> Axon.Loop.metric(:accuracy)
+      |> Axon.Loop.checkpoint()
+      |> Axon.Loop.early_stop("accuracy")
+
+  That will ensure checkpoint is always fired, even if the loop
+  exited early.
+  """
+  def early_stop(%Loop{} = loop, monitor, opts \\ []) do
+    event = opts[:event] || :epoch_completed
+    filter = opts[:filter] || :always
+    patience = opts[:patience] || 3
+    mode = opts[:mode] || :min
+
+    early_stop_fn = fn %State{metrics: metrics, handler_metadata: handler_meta} = state ->
+      unless Map.has_key?(metrics, monitor) do
+        raise ArgumentError,
+              "invalid metric to monitor, key #{inspect(monitor)} not present in metrics"
+      end
+
+      cur_criteria_value = metrics[monitor]
+
+      {prev_criteria_value, since_last_improvement} =
+        case handler_meta[:early_stop] do
+          nil ->
+            nil
```

Comment:
Will not this be a match error for `nil`?

Reply:
Good catch!